File: FRONT-END/src/api/sync.ts
import { apiClient } from './client';

export interface SyncOperationItemPayload {
  operation_id: string;
  entity_type: string;
  entity_id: string;
  operation: 'CREATE' | 'UPDATE' | 'DELETE';
  timestamp: string;
  data: Record<string, any>;
}

export interface SyncPushPayload {
  device_id: string;
  patient_id: string;
  platform?: string;
  app_version?: string;
  operations: SyncOperationItemPayload[];
}

export interface SyncPushBackendResponse {
  processed_count: number;
  successful_op_ids: string[];
  failed_ops: Array<Record<string, any>>;
  server_time: string;
}

export interface SyncPullPayload {
  patient_id: string;
  device_id: string;
  last_synced_at?: string;
}

export interface SyncPullBackendResponse {
  patient_id: string;
  server_time: string;
  games: Array<Record<string, any>>;
  active_medications: Array<Record<string, any>>;
  active_reminders: Array<Record<string, any>>;
  family_members: Array<Record<string, any>>;
  daily_activities: Array<Record<string, any>>;
  cultural_assets: Array<Record<string, any>>;
}

export const syncApi = {
  async pushBatch(payload: SyncPushPayload): Promise<SyncPushBackendResponse> {
    return apiClient<SyncPushBackendResponse>('/sync/push', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },

  async pullDelta(payload: SyncPullPayload): Promise<SyncPullBackendResponse> {
    return apiClient<SyncPullBackendResponse>('/sync/pull', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },
};

File: Backend/app/schemas/sync.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

from app.utils.enums import SyncOperationType


class SyncOperationItem(BaseModel):
    operation_id: str = Field(..., description="Client unique UUID for idempotent operation tracking")
    entity_type: str = Field(..., description="e.g. GAME_SESSION, GAME_RESULT, REMINDER_LOG, PATIENT_PREFS")
    entity_id: str = Field(..., description="Target entity UUID")
    operation: SyncOperationType = Field(default=SyncOperationType.CREATE)
    timestamp: datetime = Field(..., description="Time operation occurred offline on client device")
    data: Dict[str, Any] = Field(..., description="Entity payload data")


class SyncPushRequest(BaseModel):
    device_id: str = Field(..., description="Client device hardware/app UUID")
    patient_id: str = Field(..., description="Patient UUID")
    platform: Optional[str] = "android"
    app_version: Optional[str] = "1.0.0"
    operations: List[SyncOperationItem] = Field(default_factory=list)


class SyncPushResponse(BaseModel):
    processed_count: int
    successful_op_ids: List[str]
    failed_ops: List[Dict[str, Any]]
    server_time: datetime


class SyncPullRequest(BaseModel):
    patient_id: str
    device_id: str
    last_synced_at: Optional[datetime] = None


class SyncPullResponse(BaseModel):
    patient_id: str
    server_time: datetime
    games: List[Dict[str, Any]]
    active_medications: List[Dict[str, Any]]
    active_reminders: List[Dict[str, Any]]
    family_members: List[Dict[str, Any]]
    daily_activities: List[Dict[str, Any]]
    cultural_assets: List[Dict[str, Any]]
